File: mobile/lib/ui/features/driving/views/pedal_panel.dart
import 'package:flutter/material.dart';

import '../../../../data/services/pedal_input.dart';

/// Returns the pedal's hue: blue accelerator, red brake, yellow clutch.
Color _hue(PedalType pedal) => switch (pedal) {
  PedalType.accelerator => const Color(0xFF1E88E5),
  PedalType.brake => const Color(0xFFE53935),
  PedalType.clutch => const Color(0xFFFDD835),
};

/// Three vertical draggable pedal bars: accelerator, brake, and clutch.
///
/// Each bar maps touch position to analog pressure (0.0 at the top, 1.0 at the
/// bottom) and drives [PedalInput] while dragging, releasing to spring back on
/// drag end. Bars are hue-coded instead of labeled.
class PedalPanel extends StatelessWidget {
  const PedalPanel({super.key, required this.input, this.layout});

  final PedalInput input;
  final List<PedalType>? layout;

  @override
  Widget build(BuildContext context) {
    final order =
        layout ??
        const [PedalType.clutch, PedalType.brake, PedalType.accelerator];
    return Row(
      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
      children: order.map((pedal) {
        final keyName = switch (pedal) {
          PedalType.clutch => 'pedal-clutch',
          PedalType.brake => 'pedal-brake',
          PedalType.accelerator => 'pedal-accelerator',
        };
        return PedalBar(
          key: ValueKey(keyName),
          pedal: pedal,
          pressure: input.pressureOf(pedal),
          onDrag: input.setPressure,
          onRelease: input.release,
        );
      }).toList(),
    );
  }
}

/// A single vertical pedal bar. The filled portion grows downward from the top
/// as [pressure] increases.
class PedalBar extends StatefulWidget {
  const PedalBar({
    super.key,
    required this.pedal,
    required this.pressure,
    required this.onDrag,
    required this.onRelease,
    this.width = 64,
  });

  final PedalType pedal;
  final double pressure;
  final void Function(PedalType pedal, double pressure) onDrag;
  final void Function(PedalType pedal) onRelease;

  /// Bar width. Null lets the bar fill the width its parent gives it, which
  /// is how slot-sized grid renderers place it.
  final double? width;

  @override
  State<PedalBar> createState() => _PedalBarState();
}

class _PedalBarState extends State<PedalBar> {
  double _dragPressure = 0.0;
  bool _dragging = false;

  @override
  Widget build(BuildContext context) {
    final pressure = _dragging ? _dragPressure : widget.pressure;
    final hue = _hue(widget.pedal);

    final Widget bar = Column(
      children: [
        Expanded(
          child: LayoutBuilder(
            builder: (context, constraints) {
              final height = constraints.maxHeight;

              return GestureDetector(
                onVerticalDragStart: (details) =>
                    _setFromPosition(details.localPosition, height),
                onVerticalDragUpdate: (details) =>
                    _setFromPosition(details.localPosition, height),
                onVerticalDragEnd: (_) {
                  setState(() => _dragging = false);
                  widget.onRelease(widget.pedal);
                },
                child: Stack(
                  children: [
                    // Dark base with the pedal's hue tinted at 25% on top.
                    Container(
                      decoration: const BoxDecoration(
                        color: Color(0xFF2A2A2A),
                        borderRadius: BorderRadius.all(Radius.circular(10)),
                      ),
                    ),
                    Container(
                      decoration: BoxDecoration(
                        color: hue.withValues(alpha: 0.25),
                        borderRadius: const BorderRadius.all(
                          Radius.circular(10),
                        ),
                      ),
                    ),
                    Align(
                      alignment: Alignment.bottomCenter,
                      child: FractionallySizedBox(
                        heightFactor: pressure.clamp(0.0, 1.0),
                        child: Container(
                          decoration: BoxDecoration(
                            color: hue,
                            borderRadius: const BorderRadius.all(
                              Radius.circular(10),
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              );
            },
          ),
        ),
      ],
    );

    final width = widget.width;
    return width == null ? bar : SizedBox(width: width, child: bar);
  }

  void _setFromPosition(Offset localPosition, double height) {
    final localY = localPosition.dy;
    final pressure = (localY / height).clamp(0.0, 1.0).toDouble();
    setState(() {
      _dragging = true;
      _dragPressure = pressure;
    });
    widget.onDrag(widget.pedal, pressure);
  }
}

File: android/app/src/main/kotlin/dev/fazrigading/wheeldeck/ui/features/driving/views/PedalPanel.kt
package dev.fazrigading.wheeldeck.ui.features.driving.views

import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectVerticalDragGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import dev.fazrigading.wheeldeck.data.services.PedalInput
import dev.fazrigading.wheeldeck.domain.models.PedalType

/// Analog pressure for a touch [offsetY] down a bar [height] tall: 0.0 at the
/// top, 1.0 at the bottom, clamped outside the bar.
fun pedalPressureFromOffset(offsetY: Float, height: Float): Double =
    if (height <= 0f) 0.0 else (offsetY / height).coerceIn(0f, 1f).toDouble()

/// Applies a drag touch to a pedal: maps it to pressure and reports it. Shared
/// by the bar's drag start and its updates, and by the grid's slot-sized
/// renderers.
fun dragPedalTo(
    pedal: PedalType,
    offsetY: Float,
    height: Float,
    onDrag: (PedalType, Double) -> Unit,
): Double = pedalPressureFromOffset(offsetY, height).also { onDrag(pedal, it) }

/// The pedal's hue: blue accelerator, red brake, yellow clutch. Bars are
/// hue-coded instead of labeled.
fun pedalColor(pedal: PedalType): Color = when (pedal) {
    PedalType.Accelerator -> Color(0xFF1E88E5)
    PedalType.Brake -> Color(0xFFE53935)
    PedalType.Clutch -> Color(0xFFFDD835)
}

/// Default bar order, left to right.
val defaultPedalOrder = listOf(PedalType.Clutch, PedalType.Brake, PedalType.Accelerator)

/// Three vertical draggable pedal bars driving [input]. The filled portion grows
/// downward from the top as pressure increases; release springs back.
@Composable
fun PedalPanel(
    input: PedalInput,
    modifier: Modifier = Modifier,
    layout: List<PedalType> = defaultPedalOrder,
) {
    val pedals by input.state.collectAsState()
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceEvenly,
    ) {
        layout.forEach { pedal ->
            PedalBar(
                pedal = pedal,
                pressure = pedals[pedal],
                onDrag = input::setPressure,
                onRelease = input::release,
                modifier = Modifier.weight(1f),
            )
        }
    }
}

/// A single vertical pedal bar.
@Composable
fun PedalBar(
    pedal: PedalType,
    pressure: Double,
    onDrag: (PedalType, Double) -> Unit,
    onRelease: (PedalType) -> Unit,
    modifier: Modifier = Modifier,
) {
    val hue = pedalColor(pedal)
    // While dragging, the finger owns the fill: the spring-back animation must
    // not fight the gesture.
    var dragPressure by remember { mutableStateOf<Double?>(null) }
    val shown = (dragPressure ?: pressure).coerceIn(0.0, 1.0)

    Box(
        modifier = modifier
            .fillMaxHeight()
            .clip(RoundedCornerShape(10.dp))
            .background(Color(0xFF2A2A2A))
            .background(hue.copy(alpha = 0.25f))
            .semantics { contentDescription = "Pedal ${pedal.name}" }
            .pointerInput(pedal) {
                val barHeight = size.height.toFloat()
                detectVerticalDragGestures(
                    onDragStart = { offset ->
                        dragPressure = dragPedalTo(pedal, offset.y, barHeight, onDrag)
                    },
                    onVerticalDrag = { change, _ ->
                        change.consume()
                        dragPressure = dragPedalTo(pedal, change.position.y, barHeight, onDrag)
                    },
                    onDragEnd = {
                        dragPressure = null
                        onRelease(pedal)
                    },
                    onDragCancel = {
                        dragPressure = null
                        onRelease(pedal)
                    },
                )
            },
    ) {
        Box(
            Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
                .fillMaxHeight(shown.toFloat())
                .background(hue),
        )
    }
}
